File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        self.diameter = 0 # Global varibale to keep track of maximum diameter

        #Function to visit each node recursively and clacualte left and right depth and return max depth from that level
        def depth(root):

            if root == None:
                return 0
            
            left_depth = depth(root.left)
            right_depth = depth(root.right)

            self.diameter = max(self.diameter, left_depth + right_depth)

            return max(left_depth, right_depth) + 1
        
        depth(root)

        return self.diameter

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    // function to compute height of the binary tree
    int height(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        
        // calculate the left height of the bin tree

        int lheight = height(root->left);
        int rheight = height(root->right);

        int height = max(lheight, rheight) +1;

        return height;
    
    }

    int diameterOfBinaryTree(TreeNode* root) {

        if(root==nullptr){
            return 0;
        }

        // find height of the left subtree
        int lheight = height(root->left);

        // find height of the right subtree
        int rheight = height(root->right);

        // Get the diameter of left and right sub-trees
        int ldiameter = diameterOfBinaryTree(root->left);
        int rdiameter = diameterOfBinaryTree(root->right);

        // max depth: Max of (lheight + rheight), ldiameter, rdiameter

        return max({(lheight + rheight), ldiameter, rdiameter});
            
    }
};